lib/krb5/expand_path.c

```diff
@@ -449,11 +476,20 @@ _expand_token(krb5_context context,
 	    return _expand_extra_token(context, p[1], ret);
     }
 
-    for (i = 0; i < sizeof(tokens)/sizeof(tokens[0]); i++) {
-	if (!strncmp(token+2, tokens[i].tok, (token_end - token) - 2))
-	    return tokens[i].exp_func(context, tokens[i].param,
-				      tokens[i].postfix, ret);
-    }
+    colon = strchr(token+2, ':');
+    if (!colon)
+	colon = token_end;
+
+    for (i = 0; i < sizeof(tokens)/sizeof(tokens[0]); i++)
+	if (!strncmp(token+2, tokens[i].tok, (colon - token) - 2)) {
+	    char *arg = NULL;
+	    if (*colon == ':')
+		asprintf(&arg, "%.*s", (int)(token_end - colon - 1), colon + 1);
```

Comment:
Should we check that `asprintf` succeeded?

Reply:
We "do". That is, we pass NULL to the function which is what it would expect if no :<stuff> were specified. Like the above comment, there's a question about what the correct behaviour is on error. I was assuming that we can just proceed---but it's not optimal. It's not brilliant to just fail, either. So, we have another choice to make.

Reply:
But a test with `date +"%f%J%i" ` shows that invalid conversions just emit the letter in question and drop the `%`, so I get `fJi` for the above...

Reply:
For path expansions, I'd like to fail over using some path other than intended... Which may mean we can't log, but so be it?

Reply:
Yes, any decision that we make here is suboptimal but I agree that it is better to fail to log than it is to log to the wrong file. I'll change this and the above to fail to log. In practice, this shouldn't really happen, but we should check.

Another question arises which is: should we try to ensure that the entire log path doesn't have a malloc(3) in it? This may be one of the cases where, perhaps, it would be better to use a fixed size buffer on the stack. After all, we do not want 10K single log messages in any case. Yes, that's the change that I'll make. Fixed size buffers.

Reply:
Okay, the rest of the token expansion code doesn't use fixed size buffers, so I'm not going to in this one. I do punt on strftime(3) errors being defined as either 0 or max. That said, there is an overlap between strftime(3) errors and valid but degenerate cases, but perhaps we must accept this.

Reply:
Oddly, I've seen advice that the return value *and* the result of `asprintf()` needs to be checked.  I forget what old C libraries are perverse enough to make that needed.